### PR TITLE
Image loader tool

### LINKED ISSUE #1045: Write script for image transportation
### Background
For offline purpose all images used within kubermatic need to be available in offline environments.

### Script
Therefore we need a script which consists of 2 parts
1. Downloading all images from the internet to the local machine
2. Re-tag the images so they point to a local registry & Push all images to the local registry

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

Downloads, retags and uploads all docker images we currently deploy via the clusterController

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially resolves #1045

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/cmd/image-loader/main.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+	apiv1 "github.com/kubermatic/kubermatic/api/pkg/api/v1"
+	"github.com/kubermatic/kubermatic/api/pkg/controller/cluster"
+	"github.com/kubermatic/kubermatic/api/pkg/controller/version"
+	clusterv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
+	"github.com/kubermatic/kubermatic/api/pkg/provider"
+	"github.com/kubermatic/kubermatic/api/pkg/resources"
+
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: We normally group imports:
- go-sdk
- others
- kubermatic
- k8s

File: api/cmd/image-loader/main.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+)
+
+const mockNamespaceName = "mock-namespace"
+
+func main() {
+
+	var masterResources string
+	var requestedVersion string
+	var registryName string
+	var printOnly bool
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: Should be moved out of the function and into a `var` block

File: api/cmd/image-loader/main.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+	flag.StringVar(&masterResources, "master-resources", "../config/kubermatic/static/master/versions.yaml", "")
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: Why is this flag called `master-resources` when the default points to the `versions.yaml` ?

File: api/cmd/image-loader/main.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	flag.StringVar(&requestedVersion, "version", "", "")
+	flag.StringVar(&registryName, "registry-name", "", "Name of the registry to push to")
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: Please add an example which shows the format. Like `docker.io|quay.io`

File: api/cmd/image-loader/main.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	flag.BoolVar(&printOnly, "print-only", false, "Only print the names of found images")
+	flag.Parse()
+
+	if registryName == "" && !printOnly {
+		glog.Fatalf("Registry name must not be empty!")
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: Rather write `registry-name must not be empty` So there's no confusion about what `Registry` is

File: api/cmd/image-loader/main.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+	}
+
+	versions, err := version.LoadVersions(masterResources)
+	if err != nil {
+		glog.Fatalf("Error loading versions: %v", err)
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: include the path to the file in the error message. Unless that's already in `err`

File: api/cmd/image-loader/main.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+	}
+
+	var imagesUnfiltered []string
+	if requestedVersion == "" {
+		glog.Infof("No version passed, downloading images for all available versions...")
+		for version := range versions {
+			returnedImages, err := getImagesForVersion(versions, version)
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: We should log what versions we're processing

File: api/cmd/image-loader/main.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	if err = pushImages(retaggedImages); err != nil {
+		glog.Fatalf(err.Error())
+	}
+}
+
+func pushImages(imageTagList []string) error {
+	for _, image := range imageTagList {
+		glog.Infof("Pushing image %s", image)
+		if out, err := exec.Command("docker", "push", image).CombinedOutput(); err != nil {
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: Log the command you're going to execute

Other thing:
Should we add a retry logic?

File: api/cmd/image-loader/main.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	return nil
+}
+
+func retagImages(registryName string, imageTagList []string) (retaggedImages []string, err error) {
+	for _, image := range imageTagList {
+		imageSplitted := strings.Split(image, "/")
+		retaggedImageName := fmt.Sprintf("%s/%s", registryName, strings.Join(imageSplitted[1:], "/"))
+		glog.Infof("Tagging image %s as %s", image, retaggedImageName)
+		if out, err := exec.Command("docker", "tag", image, retaggedImageName).CombinedOutput(); err != nil {
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: Log the command you're going to execute

File: api/cmd/image-loader/main.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+		}
+	}
+
+	return retaggedImages, nil
+}
+
+func downloadImages(images []string) error {
+	for _, image := range images {
+		glog.Infof("Downloading image '%s'...\n", image)
+		if out, err := exec.Command("docker", "pull", image).CombinedOutput(); err != nil {
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: Log the command you're going to execute

Other thing:
Should we add a retry logic?

File: api/cmd/image-loader/main.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+	fakeCluster.Spec.Cloud = &clusterv1.CloudSpec{}
+	fakeCluster.Status.NamespaceName = mockNamespaceName
+	fakeCluster.Address = &clusterv1.ClusterAddress{}
+
+	go configMapInformer.Informer().Run(wait.NeverStop)
+	go secretInformer.Informer().Run(wait.NeverStop)
+	go serviceInformer.Informer().Run(wait.NeverStop)
+	cache.WaitForCacheSync(wait.NeverStop, configMapInformer.Informer().HasSynced)
+	cache.WaitForCacheSync(wait.NeverStop, secretInformer.Informer().HasSynced)
+	cache.WaitForCacheSync(wait.NeverStop, serviceInformer.Informer().HasSynced)
```
Comment (mrIncompetent) [inline api/cmd/image-loader/main.go]: You can replace the above 6 lines with the kubeInformerFactory functions.
```go
kubeInformerFactory.Start()
kubeInformerFactory.WaitForCacheSync()
```

**Review by mrIncompetent — CHANGES_REQUESTED** [review]

**Comment by mrIncompetent** [comment]
Do you want to add a Readme here or are you going to open a separate PR for it?

File: api/cmd/image-loader/main.go (hunk 10 above)
Comment (metalmatze) [reply api/cmd/image-loader/main.go]: If you want to do retry logic I can highly recommend:
https://github.com/cenkalti/backoff

File: api/cmd/image-loader/main.go (hunk 3 above)
Comment (alvaroaleman) [reply api/cmd/image-loader/main.go]: The flagname is taken over from the cluster-controller cmd

File: api/cmd/image-loader/main.go (hunk 8 above)
Comment (alvaroaleman) [reply api/cmd/image-loader/main.go]: No, I'd argue the retry logic will be the user executing this tool again

**Comment by alvaroaleman** [comment]
Regarding the readme I thought this would be a good candidate for our new hugo doc system, allhought it obviously makes sense to version it together with the code - We could probably keep it in both places for the time being and at some future point write some exporting logic that exports docs that are versioned here into the docs repo?

**Comment by mrIncompetent** [comment]
Regarding the docs:
sgtm

@cbrgm where do we have to write our documentation so it gets published via Hugo?
Also where is the documentation available?

File: api/cmd/image-loader/main.go (hunk 3 above)
Comment (mrIncompetent) [reply api/cmd/image-loader/main.go]: Right, i think by now we could actually replace that flag by dedicated version&update flags.
Though something for another PR

File: api/cmd/image-loader/main.go (hunk 8 above)
Comment (mrIncompetent) [reply api/cmd/image-loader/main.go]: sgtm

**Comment by cbrgm** [comment]
Documentation is available here https://github.com/kubermatic/docs/tree/kubermatic-docs.
You simply have to drop markdown files with some annotated meta data on top of the file into the `/content` folder. After that you can run the `hugo` command and it will generate all static website files into a folder called `/public`.

File: api/cmd/image-loader/main.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+	return &resources.TemplateData{Version: version,
```
Comment (p0lyn0mial) [inline api/cmd/image-loader/main.go]: how do we make use of `Version` ?
I think that `resources.TemplateData` doesn't have `Version` member.

File: api/cmd/image-loader/main.go [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+		if out, err := exec.Command("docker", "push", image).CombinedOutput(); err != nil {
+			return fmt.Errorf("Failed to push image: Error: %v Output: %s", err, string(out))
+		}
+	}
+	return nil
+}
+
+func retagImages(registryName string, imageTagList []string) (retaggedImages []string, err error) {
+	for _, image := range imageTagList {
+		imageSplitted := strings.Split(image, "/")
```
Comment (p0lyn0mial) [inline api/cmd/image-loader/main.go:112]: why not to check the length of `imageSplitted` ?

File: api/cmd/image-loader/main.go [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+		Items: []corev1.Service{apiServerExternalService},
+	}
+	secretList := createNamedSecrets([]string{"ca-cert",
+		"ca-key",
+		"tokens",
+		"apiserver-tls",
+		"kubelet-client-certificates",
+		"service-account-key"})
+	objects := []runtime.Object{configMapList, secretList, serviceList}
+	client := kubefake.NewSimpleClientset(objects...)
```
Comment (p0lyn0mial) [inline api/cmd/image-loader/main.go:241]: nice :)

File: api/cmd/image-loader/main.go (hunk 12 above)
Comment (mrIncompetent) [reply api/cmd/image-loader/main.go]: Not anymore. The controller exclusively takes the version from the Cluster object.
The version/update handling got some polishing in https://github.com/kubermatic/kubermatic/pull/1095

File: api/cmd/image-loader/main.go (hunk 13 above)
Comment (alvaroaleman) [reply api/cmd/image-loader/main.go:112]: Good point, added it

**Comment by alvaroaleman** [comment]
@mrIncompetent @p0lyn0mial I think I addressed all comments, PTAL. Had to update some of the version loading logic because it got significantly changed in #1095

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**